docs/refman/ppr-lprm.1.pod
#
# mouse:~ppr/src/docs/refman/ppr-lprm.1.pod
# Copyright 1995--2005, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 14 January 2005.
#

=head1 NAME

ppr-lprm - Deletes print jobs

=head1 SYNOPSIS

B<ppr-lprm> [B<-P> I<queue>] [B<->] [B<job number>] [B<user>]

=head1 DESCRIPTION

This program is part of the PPR suite.

The program B<ppr-lprm> is used to delete print jobs.  On your system,
the B<lprm> command may be an alias for B<ppr-lprm>.

There are several ways to indicate which queue one wants to remove a job from.
If the B<-P> switch is used, it overrides all other settings.  If the B<-P>
switch is not used, then B<ppr-lprm> falls back on the PRINTER environment
variable.  If the PRINTER environment variable is not defined, 
then B<ppr-lprm> assumes a queue name of "default".

There are three ways to specify which jobs within the selected queue should
be deleted.

If one or more job id numbers are specified, then the jobs with those id
numbers will be deleted.

If one or more user names are specified, then all jobs owned by those users
will be deleted.  Since ordinary users may not delete other people's jobs,
this option is only useful to the user "root".  (Note that PPR operators cannot
delete other people's jobs with B<ppr-lprm> as they can with B<ppop>.)

If the B<-> option is used, then all job belonging to the user who is running
B<ppr-lprm> are deleted.  If the user "root" is the one running B<ppr-lprm>, then all
jobs in the queue are deleted.

Note that if the remote server is running PPR's B<lprsrv>, then the definition
of "all jobs owned by this user" or "all jobs in the queue" may be limited
to those jobs submitted from the network node on which B<ppr-lprm> is being run.
However, the "proxy class=" setting in the F<lprsrv.conf> file can be used to
expand this scope at least somewhat.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut



History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

ppr-bsd/ppr-lprm.c
/*
** mouse:~ppr/src/ppr-bsd/ppr-lprm.c
** Copyright 1995--2005, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 14 January 2005.
*/

#include "config.h"
#include <unistd.h>
#include <stdlib.h>
#include <stdarg.h>
#include <pwd.h>
#include <string.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "uprint.h"

extern char *optarg;
extern int optind;

static const char *const myname = "ppr-lprm";

void uprint_error_callback(const char *format, ...)
	{
	va_list va;
	fprintf(stderr, "%s: ", myname);
	va_start(va, format);
	vfprintf(stderr, format, va);
	fputc('\n', stderr);
	va_end(va);
	} /* end of uprint_error_callback() */

static const char *option_list = "P:";

int main(int argc, char *argv[])
	{
	uid_t uid;
	struct passwd *pw;
	int c;
	char *user;
	const char *queue = (const char *)NULL;

	/* Initialize internation messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/* Trap loops: */
	if(uprint_loop_check(myname) == -1)
		return 1;

	/* We need this to look up the user name, plus we 
	   behave differently if running as root. */
	uid = getuid();

	/*
	** Look up the name of the user who invoked this program
	** and save it in a variable for future reference.
	*/
	if((pw = getpwuid(uid)) == (struct passwd *)NULL)
		{
		fprintf(stderr, _("%s: getpwuid(%ld) failed to find your account\n"), myname, (long int)uid);
		return 1;
		}
	if(pw->pw_name == (char*)NULL)
		{
		fprintf(stderr, "%s: strange getpwuid() error, pw_name is NULL\n", myname);
		return 1;
		}
	user = pw->pw_name;

	/*
	** Parse the switches.  Mostly, we will call uprint
	** member functions.
	*/
	while((c = getopt(argc, argv, option_list)) != -1)
		{
		switch(c)
			{
			case 'P':			/* printer */
				queue = optarg;
				break;

			default:
				/* fprintf(stderr, _("%s: Syntax error, unrecognized switch: -%c\n"), myname, c); */
				return 1;
			}
		}

	/* If the print destination has not yet been
	   determined, determine it now. */
	if(!queue)
		queue = uprint_default_destinations_lpr();

	/* Handle that funny stuff with an argument of "-": */
	if(argc > optind && strcmp(argv[optind], "-") == 0)
		{
		/* If root, use special incantation for all jobs of all users: */
		if(uid == 0)
			{
			argv[optind] = (char*)NULL;
			user = "-all";
			}
		/* If not root, "-" is an alias for the user name: */
		else
			{
			argv[optind] = user;
			}
		}

	/*
	** Delete the jobs
	*/
	{
	int return_code;

	if((return_code = uprint_lprm(-1, -1, user, (const char *)NULL, queue, (const char **)&argv[optind], TRUE)) >= 0)
		{
		/* Child ran, nothing more to do. */
		}

	/* Unclaimed queue */
	else if(uprint_errno == UPE_UNDEST)
		{
		fprintf(stderr, _("%s: Print queue \"%s\" not found.\n"), myname, queue);
		/* This is the exit code that BSD lprm uses for unknown queue: */
		return_code = 1;
		}

	/* Other failure: */
	else
		{
		/* An arbitrary value intended to indicated a major failure: */
		return_code = 255;
		}

	return return_code;
	}

	} /* end of main() */

/* end of file */
